File: LC410.cpp
class Solution {
public:
    int splitArray(vector<int>& nums, int k) {
        int l = 0;
        int h = 0;
        
        for(auto i : nums){
            h += i;
            l = max(l,i);
        }

        while(l < h){
            int m = l + (h-l)/2;
            int s = 0, p = 1;

            for(auto i : nums){
                if(s + i > m){
                    s = i;
                    p++;
                }
                else
                    s += i;
            }

            if(p <= k)
                h = m;
            else
                l = m + 1;
        }

        return h;
    }
};

File: LC410.py
class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        l = max(nums)
        h = sum(nums)

        while l < h:
            m = (l + h)//2
            s = 0
            p = 1

            for i in nums:
                if s + i > m:
                    s = i
                    p += 1
                else:
                    s += i
            
            if p > k:
                l = m + 1
            else:
                h = m
            
        return h
